clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -23,122 +23,397 @@ package common
+		options.LabelSelector = w.labelSelector
+	}
+	listerWatcher := cache.NewFilteredListWatchFromClient(c.CoreV1().RESTClient(), w.kind+"s", w.namespace, optionsModifier)
+	obj, err := getObjectForKind(w.kind)
 	if err != nil {
-		return summaries, err
+		return err
+	}
+	w.isRunning = true
+	w.stopCh = make(chan struct{})
+	w.informer = cache.NewSharedInformer(listerWatcher, obj, 0)
+	w.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			go func() {
+				runtimeObj, ok := obj.(runtime.Object)
+				if !ok {
+					return
+				}
+
+				w.updateOpResourceVersion(runtimeObj)
+				w.handleObject(c, runtimeObj)
+			}()
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			go func() {
+				oldRuntimeObj, ok := oldObj.(runtime.Object)
+				if !ok {
+					return
+				}
+				newRuntimeObj, ok := newObj.(runtime.Object)
+				if !ok {
+					return
+				}
+				// Increment version regardless of change in spec.
+				if err := w.updateOpResourceVersion(newRuntimeObj); err != nil {
+					return
+				}
+				if isEqual, err := isEqualRuntimeObjectsSpec(oldRuntimeObj, newRuntimeObj); err != nil || isEqual {
+					return
+				}
+				w.handleObject(c, newRuntimeObj)
+			}()
+		},
+		DeleteFunc: func(obj interface{}) {
+			go func() {
+				runtimeObj, ok := obj.(runtime.Object)
+				if !ok {
+					return
+				}
+				if err := w.updateOpResourceVersion(runtimeObj); err != nil {
+					return
+				}
+				w.deleteObject(runtimeObj)
+			}()
+		},
+	})
+
+	go w.informer.Run(w.stopCh)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerSyncTimeout, func() {
+		close(timeoutCh)
+	})
+	if !cache.WaitForCacheSync(timeoutCh, w.informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
 	}
-	labelSelector, err := util.GetStringOrDefault(config.Params, "labelSelector", "")
+	timeoutTimer.Stop()
+
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) gather(c clientset.Interface, timeout time.Duration) error {
+	glog.Infof("%v: waiting for controlled pods measurement...", waitForControlledPodsRunningName)
+	if w.informer == nil {
+		return fmt.Errorf("metric %s has not been started", waitForControlledPodsRunningName)
+	}
+	startTime := time.Now()
+	objects, err := getRuntimeObjectsForKind(c, w.kind, w.namespace, w.labelSelector, w.fieldSelector)
 	if err != nil {
-		return summaries, err
+		return fmt.Errorf("listing objects error: %v", err)
+	}
+	desiredCount := 0
+	var maxResourceVersion uint64
+	for i := range objects {
+		runtimeObj, ok := objects[i].(runtime.Object)
+		if !ok {
+			glog.Infof("%s: cannot cast to runtime.Object: %v", objects[i])
+			continue
+		}
+		version, err := getResourceVersionFromRuntimeObject(runtimeObj)
+		if err != nil {
+			glog.Errorf("%s: retriving resource version error: %v", waitForControlledPodsRunningName, err)
+			continue
+		}
+		desiredCount++
+		if version > maxResourceVersion {
+			maxResourceVersion = version
+		}
+	}
```

Comment:
Please split that into separate function i.e. function that return maxRV from current object (list + compute)

Reply:
Done.